<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to derfinderHelper • derfinderHelper</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to derfinderHelper">
<meta property="og:description" content="derfinderHelper">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinderHelper</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.29.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/derfinderHelper.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/leekgroup/derfinderHelper/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="derfinderHelper_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to derfinderHelper</h1>
                        <h4 class="author">Leonardo Collado-Torres</h4>
            <address class="author_afil">
      Lieber Institute for Brain Development, Johns Hopkins Medical CampusCenter for Computational Biology, Johns Hopkins University<br><a class="author_email" href="mailto:#"></a><a href="mailto:lcolladotor@gmail.com" class="email">lcolladotor@gmail.com</a>
      </address>
                  
            <h4 class="date">23 November 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/leekgroup/derfinderHelper/blob/master/vignettes/derfinderHelper.Rmd"><code>vignettes/derfinderHelper.Rmd</code></a></small>
      <div class="hidden name"><code>derfinderHelper.Rmd</code></div>

    </div>

    
    
<div id="basics" class="section level1">
<h1 class="hasAnchor">
<a href="#basics" class="anchor"></a>Basics</h1>
<div id="install-derfinderhelper" class="section level2">
<h2 class="hasAnchor">
<a href="#install-derfinderhelper" class="anchor"></a>Install <em><a href="https://bioconductor.org/packages/3.14/derfinderHelper">derfinderHelper</a></em>
</h2>
<p><code>R</code> is an open-source statistical environment which can be easily modified to enhance its functionality via packages. <em><a href="https://bioconductor.org/packages/3.14/derfinderHelper">derfinderHelper</a></em> is a <code>R</code> package available via the <a href="http://bioconductor/packages/derfinderHelper">Bioconductor</a> repository for packages. <code>R</code> can be installed on any operating system from <a href="https://cran.r-project.org/">CRAN</a> after which you can install <em><a href="https://bioconductor.org/packages/3.14/derfinderHelper">derfinderHelper</a></em> by using the following commands in your <code>R</code> session:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
      <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
  <span class="op">}</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"derfinderHelper"</span><span class="op">)</span>

<span class="co">## Check that you have a valid Bioconductor installation</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/valid.html">valid</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div id="required-knowledge" class="section level2">
<h2 class="hasAnchor">
<a href="#required-knowledge" class="anchor"></a>Required knowledge</h2>
<p><em><a href="https://bioconductor.org/packages/3.14/derfinderHelper">derfinderHelper</a></em> is based on many other packages and in particular in those that have implemented the infrastructure needed for dealing with RNA-seq data. A <em><a href="https://bioconductor.org/packages/3.14/derfinderHelper">derfinderHelper</a></em> user is not expected to deal with those packages directly but will need to be familiar with <em><a href="https://bioconductor.org/packages/3.14/derfinder">derfinder</a></em>.</p>
<p>If you are asking yourself the question “Where do I start using Bioconductor?” you might be interested in <a href="http://lcolladotor.github.io/2014/10/16/startbioc/#.VkOKbq6rRuU">this blog post</a>.</p>
</div>
<div id="asking-for-help" class="section level2">
<h2 class="hasAnchor">
<a href="#asking-for-help" class="anchor"></a>Asking for help</h2>
<p>As package developers, we try to explain clearly how to use our packages and in which order to use the functions. But <code>R</code> and <code>Bioconductor</code> have a steep learning curve so it is critical to learn where to ask for help. The blog post quoted above mentions some but we would like to highlight the <a href="https://support.bioconductor.org/">Bioconductor support site</a> as the main resource for getting help: remember to use the <code>derfinder</code> or <code>derfinderHelper</code> tags and check <a href="https://support.bioconductor.org/t/derfinder/">the older posts</a>. Other alternatives are available such as creating GitHub issues and tweeting. However, please note that if you want to receive help you should adhere to the <a href="http://www.bioconductor.org/help/support/posting-guide/">posting guidelines</a>. It is particularly critical that you provide a small reproducible example and your session information so package developers can track down the source of the error.</p>
</div>
<div id="citing-derfinderhelper" class="section level2">
<h2 class="hasAnchor">
<a href="#citing-derfinderhelper" class="anchor"></a>Citing <em><a href="https://bioconductor.org/packages/3.14/derfinderHelper">derfinderHelper</a></em>
</h2>
<p>We hope that <em><a href="https://bioconductor.org/packages/3.14/derfinderHelper">derfinderHelper</a></em> will be useful for your research. Please use the following information to cite the package and the overall approach. Thank you!</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Citation info</span>
<span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span><span class="op">(</span><span class="st">"derfinderHelper"</span><span class="op">)</span></code></pre></div>
<pre><code>## 
## Collado-Torres L, Jaffe AE, Leek JT (2017). _derfinderHelper: derfinder
## helper package_. doi: 10.18129/B9.bioc.derfinderHelper (URL:
## https://doi.org/10.18129/B9.bioc.derfinderHelper),
## https://github.com/leekgroup/derfinderHelper - R package version
## 1.29.0, &lt;URL: http://www.bioconductor.org/packages/derfinderHelper&gt;.
## 
## Collado-Torres L, Nellore A, Frazee AC, Wilks C, Love MI, Langmead B,
## Irizarry RA, Leek JT, Jaffe AE (2017). "Flexible expressed region
## analysis for RNA-seq with derfinder." _Nucl. Acids Res._. doi:
## 10.1093/nar/gkw852 (URL: https://doi.org/10.1093/nar/gkw852), &lt;URL:
## http://nar.oxfordjournals.org/content/early/2016/09/29/nar.gkw852&gt;.
## 
## To see these entries in BibTeX format, use 'print(&lt;citation&gt;,
## bibtex=TRUE)', 'toBibtex(.)', or set
## 'options(citation.bibtex.max=999)'.</code></pre>
</div>
</div>
<div id="introduction-to-derfinderhelper" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-derfinderhelper" class="anchor"></a>Introduction to <em><a href="https://bioconductor.org/packages/3.14/derfinderHelper">derfinderHelper</a></em>
</h1>
<p><em><a href="https://bioconductor.org/packages/3.14/derfinderHelper">derfinderHelper</a></em> (Collado-Torres, Jaffe, and Leek, 2017) is a small package that was created to speed up the F-statistics approach implemented in the parent package <em><a href="https://bioconductor.org/packages/3.14/derfinder">derfinder</a></em>. It contains a single function, <code><a href="../reference/fstats.apply.html">fstats.apply()</a></code>, which is used to calculate the F-statistics for a given data matrix, null and an alternative models.</p>
<p>The data is generally arranged in an matrix where the rows (<span class="math inline">\(n\)</span>) are the genomic features of interest (gene-level summaries, exon-level summaries, or base-level data) and the columns (<span class="math inline">\(m\)</span>) represent the samples. The other two main arguments for <code><a href="../reference/fstats.apply.html">fstats.apply()</a></code> are the null and alternative model matrices which are <span class="math inline">\(m \times p_0\)</span> and <span class="math inline">\(m \times p\)</span> where <span class="math inline">\(p_0\)</span> is the number of covariates in the null model and <span class="math inline">\(p\)</span> is the number of covariates in the alternative model. The models have to be nested and thus by definition <span class="math inline">\(p &gt; p_0\)</span>. The end result is a vector of F-statistics with length <span class="math inline">\(n\)</span>, which is run length encoded for memory saving purposes.</p>
<p>Other arguments of <code><a href="../reference/fstats.apply.html">fstats.apply()</a></code> are related to flow in <em><a href="https://bioconductor.org/packages/3.14/derfinder">derfinder</a></em> such as the scaling factor (<code>scalefac</code>) used, whether to subset the data (<code>index</code>), and if the data was separated into chunks and saved to disk to lower the memory load (<code>lowMemDir</code>).</p>
<p>Implementation-wise, <code>adjustF</code> is useful when the denominator of the F-statistic calculation is too small. Finally, <code>method</code> controls how will the F-statistics be calculated.</p>
<ul>
<li>
<code>Matrix</code> is the recommended option because it uses around half the memory load of <code>regular</code> and can be faster. Specially if the data was saved in this format previously by <em><a href="https://bioconductor.org/packages/3.14/derfinder">derfinder</a></em>.</li>
<li>
<code>Rle</code> uses the least amount of memory but gets very slow as the number of samples increases. Thus making it less than ideal in several cases.</li>
<li>
<code>regular</code> uses base <code>R</code> to calculate the F-statistics and can require a large amount of memory. This is noticeable when using several cores to run <code><a href="../reference/fstats.apply.html">fstats.apply()</a></code> on different portions of the data.</li>
</ul>
<p>The F-statistics for each feature <span class="math inline">\(i\)</span> are calculated using the following formula:</p>
<p><span class="math display">\[ F_i = \frac{ (\text{RSS0}_i - \text{RSS1}_i)/(\text{df}_1 - \text{df}_0) }{ \text{adjustF} + (\text{RSS1}_i / (p - p_0 - \text{df_1}))} \]</span></p>
</div>
<div id="example" class="section level1">
<h1 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h1>
<p>The following section walks through an example. However, in practice, you will probably not use this package directly and it will be used via <em><a href="https://bioconductor.org/packages/3.14/derfinder">derfinder</a></em>.</p>
<div id="data" class="section level2">
<h2 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h2>
<p>First lets create an example data set where we have information for 1000 features and 16 samples where samples 1 to 4 are from group A, 5 to 8 from group B, 9 to 12 from group C, and 13 to 16 from group D.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Create some toy data</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/IRanges">"IRanges"</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">20140923</span><span class="op">)</span>
<span class="va">toyData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html">DataFrame</a></span><span class="op">(</span>
    <span class="st">"sample1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample2"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample3"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample4"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample5"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">15</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample6"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">15</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample7"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">15</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample8"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">15</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample9"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample10"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample11"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample12"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample13"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample14"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample15"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    <span class="st">"sample16"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">1000</span>, <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span>

<span class="co">## Lets say that we have 4 groups</span>
<span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/chartr.html">toupper</a></span><span class="op">(</span><span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Note that some groups have higher coverage, we can adjust for this in the model</span>
<span class="va">sampleDepth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html">sapply</a></span><span class="op">(</span><span class="va">toyData</span>, <span class="va">sum</span><span class="op">)</span>
<span class="va">sampleDepth</span></code></pre></div>
<pre><code>##  sample1  sample2  sample3  sample4  sample5  sample6  sample7  sample8 
##     4948     5044     5054     5027     7464     7407     7504     7607 
##  sample9 sample10 sample11 sample12 sample13 sample14 sample15 sample16 
##     9797     9982    10028     9797    50682    49921    49332    50421</code></pre>
</div>
<div id="models" class="section level2">
<h2 class="hasAnchor">
<a href="#models" class="anchor"></a>Models</h2>
<p>Next we create the model matrices for our example data set. Lets say that we want to calculate F-statistics comparing the alternative hypothesis that the group coefficients are not 0 versus the null hypothesis that they are equal to 0, when adjusting for the sample depth.</p>
<p>To do so, we create the nested models.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Build the model matrices</span>
<span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">sampleDepth</span> <span class="op">+</span> <span class="va">group</span><span class="op">)</span>
<span class="va">mod0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">sampleDepth</span><span class="op">)</span>

<span class="co">## Explore them</span>
<span class="va">mod</span></code></pre></div>
<pre><code>##    (Intercept) sampleDepth groupB groupC groupD
## 1            1        4948      0      0      0
## 2            1        5044      0      0      0
## 3            1        5054      0      0      0
## 4            1        5027      0      0      0
## 5            1        7464      1      0      0
## 6            1        7407      1      0      0
## 7            1        7504      1      0      0
## 8            1        7607      1      0      0
## 9            1        9797      0      1      0
## 10           1        9982      0      1      0
## 11           1       10028      0      1      0
## 12           1        9797      0      1      0
## 13           1       50682      0      0      1
## 14           1       49921      0      0      1
## 15           1       49332      0      0      1
## 16           1       50421      0      0      1
## attr(,"assign")
## [1] 0 1 2 2 2
## attr(,"contrasts")
## attr(,"contrasts")$group
## [1] "contr.treatment"</code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mod0</span></code></pre></div>
<pre><code>##    (Intercept) sampleDepth
## 1            1        4948
## 2            1        5044
## 3            1        5054
## 4            1        5027
## 5            1        7464
## 6            1        7407
## 7            1        7504
## 8            1        7607
## 9            1        9797
## 10           1        9982
## 11           1       10028
## 12           1        9797
## 13           1       50682
## 14           1       49921
## 15           1       49332
## 16           1       50421
## attr(,"assign")
## [1] 0 1</code></pre>
</div>
<div id="get-f-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#get-f-statistics" class="anchor"></a>Get F-statistics</h2>
<p>Finally, we can calculate the F-statistics using <code><a href="../reference/fstats.apply.html">fstats.apply()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/leekgroup/derfinderHelper">"derfinderHelper"</a></span><span class="op">)</span>
<span class="va">fstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fstats.apply.html">fstats.apply</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">toyData</span>, mod <span class="op">=</span> <span class="va">mod</span>, mod0 <span class="op">=</span> <span class="va">mod0</span>, scalefac <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">fstats</span></code></pre></div>
<pre><code>## numeric-Rle of length 1000 with 1000 runs
##   Lengths:          1          1          1 ...          1          1
##   Values :  1.8484614  0.5732806  0.2493710 ...  0.1197002  1.3113647</code></pre>
<p>We can then proceed to use this information in <em><a href="https://bioconductor.org/packages/3.14/derfinder">derfinder</a></em> or in any way you like.</p>
</div>
</div>
<div id="details" class="section level1">
<h1 class="hasAnchor">
<a href="#details" class="anchor"></a>Details</h1>
<p>We created <em><a href="https://bioconductor.org/packages/3.14/derfinderHelper">derfinderHelper</a></em> for calculating F-statistics using <code>SnowParam()</code> from <em><a href="https://bioconductor.org/packages/3.14/BiocParallel">BiocParallel</a></em>. Using this form of parallelization requires loading the necessary packages in the child processes. Because <em><a href="https://bioconductor.org/packages/3.14/derfinder">derfinder</a></em> takes a long time to load, we shipped off <code><a href="../reference/fstats.apply.html">fstats.apply()</a></code> to its own package to improve the speed of the calculations while retaining the memory advantages of <code>SnowParam()</code> over <code>MulticoreParam()</code>.</p>
<p>Note that transforming the data from a <code>DataFrame</code> to a <code>dgCMatrix</code> takes some time, so the most efficient performance is achieved when the data is converted at the beginning instead of at every permutation calculation. This is done in <code>derfinder::preprocessCoverage()</code> when <code>lowMemDir</code> is specified.</p>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>This package was made possible thanks to:</p>
<ul>
<li>R (R Core Team, 2021)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.14/IRanges">IRanges</a></em> (Lawrence, Huber, Pagès, Aboyoun, Carlson, Gentleman, Morgan, and Carey, 2013)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=Matrix">Matrix</a></em> (Bates and Maechler, 2021)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.14/S4Vectors">S4Vectors</a></em> (Pagès, Lawrence, and Aboyoun, 2017)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=sessioninfo">sessioninfo</a></em> (Wickham, Chang, Flight, Müller, and Hester, 2021)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> (Xie, 2014)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.14/BiocStyle">BiocStyle</a></em> (Oleś, 2021)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=RefManageR">RefManageR</a></em> (McLean, 2017)</li>
<li><em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em></li>
<li>
<em><a href="https://CRAN.R-project.org/package=testthat">testthat</a></em> (Wickham, 2011)</li>
</ul>
<p>Code for creating the vignette</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Create the vignette</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/rstudio/rmarkdown">"rmarkdown"</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html">render</a></span><span class="op">(</span><span class="st">"derfinderHelper.Rmd"</span>, <span class="st">"BiocStyle::html_document"</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Extract the R code</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://yihui.org/knitr/">"knitr"</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/knitr/man/knit.html">knit</a></span><span class="op">(</span><span class="st">"derfinderHelper.Rmd"</span>, tangle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Date the vignette was generated.</p>
<pre><code>## [1] "2021-11-23 01:06:21 UTC"</code></pre>
<p>Wallclock time spent generating the vignette.</p>
<pre><code>## Time difference of 3.599 secs</code></pre>
<p><code>R</code> session information.</p>
<pre><code>## ─ Session info  ──────────────────────────────────────────────────────────────────────────────────────────────────────
##  hash: money bag, person gesturing OK, bullet train
## 
##  setting  value
##  version  R version 4.1.2 (2021-11-01)
##  os       Ubuntu 20.04.3 LTS
##  system   x86_64, linux-gnu
##  ui       X11
##  language (EN)
##  collate  en_US.UTF-8
##  ctype    en_US.UTF-8
##  tz       UTC
##  date     2021-11-23
##  pandoc   2.14.0.3 @ /usr/local/bin/ (via rmarkdown)
## 
## ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
##  package         * version date (UTC) lib source
##  BiocGenerics    * 0.40.0  2021-10-26 [1] Bioconductor
##  BiocManager       1.30.16 2021-06-15 [2] CRAN (R 4.1.2)
##  BiocStyle       * 2.22.0  2021-10-26 [1] Bioconductor
##  bookdown          0.24    2021-09-02 [1] RSPM (R 4.1.0)
##  cachem            1.0.6   2021-08-19 [2] RSPM (R 4.1.0)
##  cli               3.1.0   2021-10-27 [2] RSPM (R 4.1.0)
##  crayon            1.4.2   2021-10-29 [2] RSPM (R 4.1.0)
##  derfinderHelper * 1.29.0  2021-11-23 [1] Bioconductor
##  desc              1.4.0   2021-09-28 [2] RSPM (R 4.1.0)
##  digest            0.6.28  2021-09-23 [2] RSPM (R 4.1.0)
##  evaluate          0.14    2019-05-28 [2] RSPM (R 4.1.0)
##  fastmap           1.1.0   2021-01-25 [2] RSPM (R 4.1.0)
##  fs                1.5.0   2020-07-31 [2] RSPM (R 4.1.0)
##  generics          0.1.1   2021-10-25 [2] RSPM (R 4.1.0)
##  htmltools         0.5.2   2021-08-25 [1] RSPM (R 4.1.0)
##  httr              1.4.2   2020-07-20 [2] RSPM (R 4.1.0)
##  IRanges         * 2.28.0  2021-10-26 [1] Bioconductor
##  jquerylib         0.1.4   2021-04-26 [1] RSPM (R 4.1.0)
##  jsonlite          1.7.2   2020-12-09 [2] RSPM (R 4.1.0)
##  knitr             1.36    2021-09-29 [2] RSPM (R 4.1.0)
##  lattice           0.20-45 2021-09-22 [3] CRAN (R 4.1.2)
##  lubridate         1.8.0   2021-10-07 [1] RSPM (R 4.1.0)
##  magrittr          2.0.1   2020-11-17 [2] RSPM (R 4.1.0)
##  Matrix            1.3-4   2021-06-01 [3] CRAN (R 4.1.2)
##  memoise           2.0.0   2021-01-26 [2] RSPM (R 4.1.0)
##  pkgdown           1.6.1   2020-09-12 [1] RSPM (R 4.1.0)
##  plyr              1.8.6   2020-03-03 [1] RSPM (R 4.1.0)
##  R6                2.5.1   2021-08-19 [2] RSPM (R 4.1.0)
##  ragg              1.2.0   2021-10-30 [1] RSPM (R 4.1.0)
##  Rcpp              1.0.7   2021-07-07 [1] RSPM (R 4.1.0)
##  RefManageR      * 1.3.0   2020-11-13 [1] RSPM (R 4.1.0)
##  rlang             0.4.12  2021-10-18 [2] RSPM (R 4.1.0)
##  rmarkdown         2.11    2021-09-14 [1] RSPM (R 4.1.0)
##  rprojroot         2.0.2   2020-11-15 [2] RSPM (R 4.1.0)
##  S4Vectors       * 0.32.3  2021-11-21 [1] Bioconductor
##  sessioninfo     * 1.2.1   2021-11-02 [2] RSPM (R 4.1.0)
##  stringi           1.7.5   2021-10-04 [2] RSPM (R 4.1.0)
##  stringr           1.4.0   2019-02-10 [2] RSPM (R 4.1.0)
##  systemfonts       1.0.3   2021-10-13 [1] RSPM (R 4.1.0)
##  textshaping       0.3.6   2021-10-13 [1] RSPM (R 4.1.0)
##  xfun              0.28    2021-11-04 [2] RSPM (R 4.1.0)
##  xml2              1.3.2   2020-04-23 [2] RSPM (R 4.1.0)
##  yaml              2.2.1   2020-02-01 [2] RSPM (R 4.1.0)
## 
##  [1] /__w/_temp/Library
##  [2] /usr/local/lib/R/site-library
##  [3] /usr/local/lib/R/library
## 
## ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
<div id="bibliography" class="section level1">
<h1 class="hasAnchor">
<a href="#bibliography" class="anchor"></a>Bibliography</h1>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.14/BiocStyle">BiocStyle</a></em> (Oleś, 2021) with <em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> (Xie, 2014) and <em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=RefManageR">RefManageR</a></em> (McLean, 2017).</p>
<p>
<a id="bib-bates2021matrix"></a><a href="#cite-bates2021matrix">[1]</a><cite> D. Bates and M. Maechler. <em>Matrix: Sparse and Dense Matrix Classes and Methods</em>. R package version 1.3-4. 2021. URL: <a href="https://CRAN.R-project.org/package=Matrix">https://CRAN.R-project.org/package=Matrix</a>.</cite>
</p>
<p>
<a id="bib-colladotorres2017derfinderhelper"></a><a href="#cite-colladotorres2017derfinderhelper">[2]</a><cite> L. Collado-Torres, A. E. Jaffe, and J. T. Leek. <em>derfinderHelper: derfinder helper package</em>. <a href="https://github.com/leekgroup/derfinderHelper" class="uri">https://github.com/leekgroup/derfinderHelper</a> - R package version 1.29.0. 2017. DOI: <a href="https://doi.org/10.18129/B9.bioc.derfinderHelper">10.18129/B9.bioc.derfinderHelper</a>. URL: <a href="http://www.bioconductor.org/packages/derfinderHelper">http://www.bioconductor.org/packages/derfinderHelper</a>.</cite>
</p>
<p>
<a id="bib-lawrence2013software"></a><a href="#cite-lawrence2013software">[3]</a><cite> M. Lawrence, W. Huber, H. Pagès, et al. “Software for Computing and Annotating Genomic Ranges”. In: <em>PLoS Computational Biology</em> 9 (8 2013). DOI: <a href="https://doi.org/10.1371/journal.pcbi.1003118">10.1371/journal.pcbi.1003118</a>. URL: <a href="http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118%7D."></a><a href="http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118" class="uri">http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118</a>}.
</cite></p>


<p>
<a id="bib-mclean2017refmanager"></a><a href="#cite-mclean2017refmanager">[4]</a><cite> M. W. McLean. “RefManageR: Import and Manage BibTeX and BibLaTeX References in R”. In: <em>The Journal of Open Source Software</em> (2017). DOI: <a href="https://doi.org/10.21105/joss.00338">10.21105/joss.00338</a>.</cite>
</p>
<p>
<a id="bib-ole2021biocstyle"></a><a href="#cite-ole2021biocstyle">[5]</a><cite> A. Oleś. <em>BiocStyle: Standard styles for vignettes and other Bioconductor documents</em>. R package version 2.22.0. 2021. URL: <a href="https://github.com/Bioconductor/BiocStyle">https://github.com/Bioconductor/BiocStyle</a>.</cite>
</p>
<p>
<a id="bib-S4Vectors"></a><a href="#cite-S4Vectors">[6]</a><cite> H. Pagès, M. Lawrence, and P. Aboyoun. <em>S4Vectors: S4 implementation of vector-like and list-like objects</em>. 2017. DOI: <a href="https://doi.org/10.18129/B9.bioc.S4Vectors">10.18129/B9.bioc.S4Vectors</a>.</cite>
</p>
<p>
<a id="bib-2021language"></a><a href="#cite-2021language">[7]</a><cite> R Core Team. <em>R: A Language and Environment for Statistical Computing</em>. R Foundation for Statistical Computing. Vienna, Austria, 2021. URL: <a href="https://www.R-project.org/">https://www.R-project.org/</a>.</cite>
</p>
<p>
<a id="bib-wickham2011testthat"></a><a href="#cite-wickham2011testthat">[8]</a><cite> H. Wickham. “testthat: Get Started with Testing”. In: <em>The R Journal</em> 3 (2011), pp. 5–10. URL: <a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf">https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf</a>.</cite>
</p>
<p>
<a id="bib-wickham2021sessioninfo"></a><a href="#cite-wickham2021sessioninfo">[9]</a><cite> H. Wickham, W. Chang, R. Flight, et al. <em>sessioninfo: R Session Information</em>. <a href="https://github.com/r-lib/sessioninfo#readme" class="uri">https://github.com/r-lib/sessioninfo#readme</a>, <a href="https://r-lib.github.io/sessioninfo/" class="uri">https://r-lib.github.io/sessioninfo/</a>. 2021.</cite>
</p>
<p>
<a id="bib-xie2014knitr"></a><a href="#cite-xie2014knitr">[10]</a><cite> Y. Xie. “knitr: A Comprehensive Tool for Reproducible Research in R”. In: <em>Implementing Reproducible Computational Research</em>. Ed. by V. Stodden, F. Leisch and R. D. Peng. ISBN 978-1466561595. Chapman and Hall/CRC, 2014. URL: <a href="http://www.crcpress.com/product/isbn/9781466561595">http://www.crcpress.com/product/isbn/9781466561595</a>.</cite>
</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
